File: server/sms_mailbox.py
"""Pure validation/state transitions shared by the HTTP SMS command endpoints."""
import re
import time

TTL_MS = 3600000
RETENTION_MS = 172800000
STATES = {"sending", "sent", "failed", "unknown"}


def now_ms():
    return int(time.time() * 1000)


def submit(records, command, now=None):
    now = now_ms() if now is None else now
    if not isinstance(command, dict):
        return 400, None
    old = next((x for x in records if x["id"] == command.get("id")), None)
    if old:
        return (200, old) if all(old.get(k) == command.get(k) for k in ("to", "body", "created_at", "subscription_id")) else (409, None)
    created = command.get("created_at")
    if not (isinstance(command.get("id"), str) and re.fullmatch(r"[a-f0-9-]{36}", command["id"])
            and isinstance(command.get("to"), str) and re.fullmatch(r"\+?[0-9]{3,20}", command["to"])
            and isinstance(command.get("body"), str) and command["body"].strip() and len(command["body"].encode('utf-16-le')) // 2 <= 1600
            and (command.get("subscription_id") is None or (type(command["subscription_id"]) is int and 0 <= command["subscription_id"] <= 2147483647))
            and type(created) is int and now - TTL_MS < created <= now + 300000):
        return 400, None
    records[:] = [x for x in records if x["created_at"] > now - RETENTION_MS]
    if len(records) >= 100:
        return 429, None
    record = {k: command[k] for k in ("id", "to", "body", "created_at")}
    record.update(subscription_id=command.get("subscription_id"), status="queued", detail="", expires_at=created + TTL_MS)
    records.append(record)
    return 200, record


def expire(records, now=None):
    now = now_ms() if now is None else now
    for x in records:
        if x["status"] == "queued" and x["expires_at"] <= now:
            x.update(status="expired", detail="Sender did not send before the request expired")


def update(records, result):
    if not isinstance(result, dict) or result.get("status") not in STATES:
        return False
    record = next((x for x in records if x["id"] == result.get("id")), None)
    if record is None:
        return False
    if record["status"] in {"sent", "failed"}:
        return True
    if record["status"] in {"unknown", "expired"} and result["status"] not in {"sent", "failed"}:
        return True
    record.update(status=result["status"], detail=str(result.get("detail", ""))[:240])
    return True


def sim_options(value, now=None):
    if not isinstance(value, dict) or value.get("state") not in {"ready", "permission_required", "unavailable"}:
        return None
    sims = value.get("sims")
    if not isinstance(sims, list) or len(sims) > 8:
        return None
    ids = set()
    clean = []
    for sim in sims:
        if not isinstance(sim, dict) or not (type(sim.get("id")) is int and 0 <= sim["id"] <= 2147483647
                and sim["id"] not in ids and type(sim.get("slot")) is int and 0 <= sim["slot"] <= 7
                and isinstance(sim.get("name"), str) and len(sim["name"].encode("utf-16-le")) // 2 <= 80
                and isinstance(sim.get("carrier"), str) and len(sim["carrier"].encode("utf-16-le")) // 2 <= 80):
            return None
        ids.add(sim["id"])
        clean.append({k: sim[k] for k in ("id", "slot", "name", "carrier")})
    default = value.get("default_id")
    if default is not None and (type(default) is not int or default not in ids):
        return None
    if value["state"] != "ready" and sims:
        return None
    return dict(state=value["state"], sims=clean, default_id=default, updated_at=now_ms() if now is None else now)

File: server-workers/src/sms-mailbox.mjs
// SMS commands are durable and idempotent. An expired command can never be re-created.
export const SMS_TTL_MS = 60 * 60 * 1000;
export const SMS_RETENTION_MS = 48 * 60 * 60 * 1000;
export const SMS_STATES = new Set(['sending', 'sent', 'failed', 'unknown']);
export function validSms(command, now = Date.now()) {
  return command && typeof command.id === 'string' && /^[a-f0-9-]{36}$/.test(command.id) &&
    typeof command.to === 'string' && /^\+?[0-9]{3,20}$/.test(command.to) &&
    typeof command.body === 'string' && command.body.trim().length > 0 && command.body.length <= 1600 &&
    (command.subscription_id == null || (Number.isInteger(command.subscription_id) && command.subscription_id >= 0 && command.subscription_id <= 2147483647)) &&
    Number.isSafeInteger(command.created_at) && command.created_at <= now + 300000 &&
    command.created_at > now - SMS_TTL_MS;
}
export function submitSms(records, command, now = Date.now()) {
  const old = records.find(x => x.id === command?.id);
  if (old) {
    if (old.to !== command.to || old.body !== command.body || old.created_at !== command.created_at || (old.subscription_id ?? null) !== (command.subscription_id ?? null))
      return { status: 409, error: 'Request ID already used' };
    return { status: 200, records, command: old };
  }
  if (!validSms(command, now)) return { status: 400, error: 'Invalid or expired SMS request' };
  const retained = records.filter(x => x.created_at > now - SMS_RETENTION_MS);
  if (retained.length >= 100) return { status: 429, error: 'SMS queue is full' };
  const next = { id: command.id, to: command.to, body: command.body, created_at: command.created_at,
    subscription_id: command.subscription_id ?? null, status: 'queued', detail: '', expires_at: command.created_at + SMS_TTL_MS };
  return { status: 200, records: [...retained, next], command: next };
}
export function updateSms(records, result) {
  if (!result || !SMS_STATES.has(result.status)) return false;
  const record = records.find(x => x.id === result.id);
  if (!record) return false;
  // Late retries cannot regress a carrier result to "sending".
  if (['sent', 'failed'].includes(record.status)) return true;
  if (['unknown', 'expired'].includes(record.status) && !['sent', 'failed'].includes(result.status)) return true;
  record.status = result.status;
  record.detail = typeof result.detail === 'string' ? result.detail.slice(0, 240) : '';
  return true;
}
export function expireSms(records, now = Date.now()) {
  return records.map(x => x.status === 'queued' && x.expires_at <= now
    ? { ...x, status: 'expired', detail: 'Sender did not send before the request expired' } : x);
}

// Only sender-authenticated fetches may publish this small, non-identifying inventory.
export function simOptions(value, now = Date.now()) {
  if (!value || !['ready', 'permission_required', 'unavailable'].includes(value.state) ||
      !Array.isArray(value.sims) || value.sims.length > 8) return null;
  const ids = new Set();
  const sims = [];
  for (const sim of value.sims) {
    if (!sim || !Number.isInteger(sim.id) || sim.id < 0 || sim.id > 2147483647 || ids.has(sim.id) ||
        !Number.isInteger(sim.slot) || sim.slot < 0 || sim.slot > 7 ||
        typeof sim.name !== 'string' || sim.name.length > 80 || typeof sim.carrier !== 'string' || sim.carrier.length > 80) return null;
    ids.add(sim.id);
    sims.push({id: sim.id, slot: sim.slot, name: sim.name, carrier: sim.carrier});
  }
  if (value.default_id != null && !ids.has(value.default_id)) return null;
  if (value.state !== 'ready' && sims.length) return null;
  return {state: value.state, sims, default_id: value.default_id ?? null, updated_at: now};
}
